### PR TITLE
Major command line parsing overhaul (internal)

### PR DESCRIPTION
**Pull request by Tarrasch**
See commit message.

### CODE REVIEW

**Comment by Tarrasch** [comment]
@erikbern, what do you about this?

File: luigi/task.py [hunk 1]
```diff
@@ -253,18 +248,6 @@ def list_to_tuple(x):
...
-
-        For example, the Task:
-
-        .. code-block:: python
-
-            class MyTask(luigi.Task):
-                count = luigi.IntParameter()
-
-        can be instantiated as ``MyTask(count=10)``.
-        """
```
Comment (Tarrasch) [inline luigi/task.py]: This documentation was just unreadable and in the way.

File: luigi/tools/deps.py [hunk 2]
```diff
@@ -16,7 +16,7 @@
 # You fixed the bug and now you want to rerun it, including all it's upstream deps.
 #
 # To do that you run:
-#      bin/deps.py --module daily_module Aggregate --daily-param1 xxx --upstream Daily
+#      bin/deps.py --module daily_module Aggregate --daily-param1 xxx --upstream-family Daily
```
Comment (Tarrasch) [inline luigi/tools/deps.py]: Unrelated to this commit, but still fixes.

File: luigi/interface.py [hunk 3]
```diff
@@ -187,121 +188,6 @@ def _schedule_and_run(tasks, worker_scheduler_factory=None, override_defaults=No
...
-def _get_task_parameters(task_cls, args):
-    # Parse a str->str dict to the correct types
-    params = {}
-    for param_name, param in task_cls.get_params():
-        param.parse_from_args(param_name, task_cls.task_family, args, params)
-    return params
-
-
-def _set_global_parameters(args):
-    # Note that this is not side effect free
```
Comment (Tarrasch) [inline luigi/interface.py]: Die you unreadable unpure code!!! >_<

File: luigi/parameter.py [hunk 4]
```diff
@@ -176,7 +187,16 @@ def _value_iterator(self, task_name, param_name):
 
         The parameter value will be whatever non-_no_value that is yielded first.
         """
-        yield (self.__global, None)
+        from luigi.cmdline_parser import CmdlineParser
+        cp_parser = CmdlineParser.get_instance()
+        if cp_parser:
```
Comment (Tarrasch) [inline luigi/parameter.py]: @erikbern, as you hoped for in #1197, it's not hard to not force an `CmdlinePaser` instance to be present. :)

File: doc/configuration.rst [hunk 5]
```diff
@@ -32,16 +38,55 @@ configuration file. For instance if you have a Task definition:
...
 ::
 
     [DailyReport]
     date: 2012-01-01
 
-You can also use ``config_path`` as an argument to the ``Parameter`` if
-you want to use a specific section in the config.
+.. _ConfigClasses:
+
+Configuration classes
```
Comment (Tarrasch) [inline doc/configuration.rst]: Being able to hyperlink to this documentation is going to save me sooooo much time. :)

**Comment by Tarrasch** [comment]
@erikbern, can you review this?

File: test/cmdline_test.py [hunk 6]
```diff
@@ -202,21 +202,14 @@ def test_bin_luigi_help_class(self):
...
 
+    @parsing(['FooSubClass', '--x', 'xyz', '--FooBaseClass-x', 'xyz'])
     def test_subclasses(self):
-        ap = luigi.interface._ArgParseInterface()
-
-        task, = ap.parse(['--local-scheduler', '--no-lock', 'FooSubClass', '--x', 'xyz', '--FooBaseClass-x', 'xyz'])
-        self.assertEquals(task.x, 'xyz')
-
-        # This won't work because --FooSubClass-x doesn't exist
-        self.assertRaises(BaseException, ap.parse, (['--local-scheduler', '--no-lock', 'FooBaseClass', '--x', 'xyz', '--FooSubClass-x', 'xyz']))
```
Comment (Tarrasch) [inline test/cmdline_test.py]: I think you wrote this @erikbern. I had to remove this `assertRaises` as I did not understand it. It doesn't make sense to me at first sight. Can you comment?

**Comment by Tarrasch** [comment]
F yeah! it seems to be working consistently now after I rebased on top of #1220. In other words, it's prime time to review now @erikbern! :)

I ran the 33 and 34 tests 5 times each now at least and no sporadic failure that was introduced as of this patch.

(I realized there is failure in https://travis-ci.org/spotify/luigi/jobs/80641915, but that's not due to this PR, it's from a new test in #1214 being shaky)

**Comment by erikbern** [comment]
ok will try to do it this weekend! (first weekend in i think 3 months that i'm not busy)

**Comment by Tarrasch** [comment]
It's not so user friendly, but if you comment on the PR instead of the individual commit @ulzha, then comments are preserved in the PR even if I force push my amendment. Anyhow, here's the commit you commented on: https://github.com/Tarrasch/luigi/commit/92b9e885c1067c4a5ba06a84ae6e0c37370a6873

**Comment by Tarrasch** [comment]
@erikbern, @toidi did you guys have a chance to review?

File: test/cmdline_test.py (hunk 6 above)
Comment (erikbern) [inline test/cmdline_test.py]: not sure either why it would raise

**Comment by erikbern** [comment]
LGTM but there's no dependency on the command line parser right?

File: luigi/parameter.py [hunk 7]
```diff
@@ -176,7 +187,15 @@ def _value_iterator(self, task_name, param_name):
 
         The parameter value will be whatever non-_no_value that is yielded first.
         """
-        yield (self.__global, None)
+        cp_parser = CmdlineParser.get_instance()
```
Comment (erikbern) [inline luigi/parameter.py]: wait so CmdlineParser is a singleton now? not sure if i'm a big fan

**Comment by erikbern** [comment]
I don't like these two things
1. The inversion of dependency between `Parameter` and the command line parser
2. The fact that that the command line parser is a singleton

But other than that, LGTM. There's so much cruft in this code anyway that I think you're removing a lot worse stuff

**Comment by Tarrasch** [comment]
@erikbern. Well yea. I don't see a good way to get rid of globals without turning the syntax `MyTask(my_param=123)` into `worker.instantiate_task(MyTask, param=123)`. Who should hold/own the parsing context if you don't have it global.

**Comment by erikbern** [comment]
cool dude let's merge

**Comment by Tarrasch** [comment]
Awesome. Now I think we just need to fix the docs a little bit more (it should all be well-formed on reatthedocs), and then we have a shiny 2.0.0 to release!

**Comment by erikbern** [comment]
yay!

**Pull request merged**